clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java

```diff
@@ -132,4 +414,332 @@ public synchronized void incrementSequenceNumber(TopicPartition topicPartition,
+        EndTxnRequest.Builder builder = new EndTxnRequest.Builder(transactionalId,
+                pidAndEpoch.producerId, pidAndEpoch.epoch, isCommit ? TransactionResult.COMMIT : TransactionResult.ABORT);
+        return new TransactionalRequest(builder, new EndTxnCallback(isCommit, result),
+                FindCoordinatorRequest.CoordinatorType.TRANSACTION, TransactionalRequest.Priority.END_TXN, isRetry, transactionalId);
+    }
+
+    private TransactionalRequest addOffsetsToTxnRequest(Map<TopicPartition, OffsetAndMetadata> offsets,
+                                                        String consumerGroupId, boolean isRetry, TransactionalRequestResult result) {
+        AddOffsetsToTxnRequest.Builder builder = new AddOffsetsToTxnRequest.Builder(transactionalId,
+                pidAndEpoch.producerId, pidAndEpoch.epoch, consumerGroupId);
+        return new TransactionalRequest(builder, new AddOffsetsToTxnCallback(offsets, consumerGroupId, result),
+                FindCoordinatorRequest.CoordinatorType.TRANSACTION, TransactionalRequest.Priority.ADD_PARTITIONS_OR_OFFSETS, isRetry, transactionalId);
+    }
+
+    private TransactionalRequest txnOffsetCommitRequest(Map<TopicPartition, OffsetAndMetadata> offsets,
+                                                        String consumerGroupId, boolean isRetry, TransactionalRequestResult result) {
+        for (Map.Entry<TopicPartition, OffsetAndMetadata> entry : offsets.entrySet()) {
+            OffsetAndMetadata offsetAndMetadata = entry.getValue();
+            pendingTxnOffsetCommits.put(entry.getKey(),
+                    new TxnOffsetCommitRequest.CommittedOffset(offsetAndMetadata.offset(), offsetAndMetadata.metadata()));
+        }
+        return txnOffsetCommitRequest(consumerGroupId, isRetry, result);
+    }
+
+    private TransactionalRequest txnOffsetCommitRequest(String consumerGroupId, boolean isRetry, TransactionalRequestResult result) {
+        TxnOffsetCommitRequest.Builder builder = new TxnOffsetCommitRequest.Builder(consumerGroupId,
+                pidAndEpoch.producerId, pidAndEpoch.epoch, OffsetCommitRequest.DEFAULT_RETENTION_TIME, pendingTxnOffsetCommits);
+        return new TransactionalRequest(builder, new TxnOffsetCommitCallback(consumerGroupId, result),
+                FindCoordinatorRequest.CoordinatorType.GROUP, TransactionalRequest.Priority.ADD_PARTITIONS_OR_OFFSETS, isRetry, consumerGroupId);
+    }
+
+    private abstract class TransactionalRequestCallBack implements RequestCompletionHandler {
+        protected final TransactionalRequestResult result;
+
+        TransactionalRequestCallBack(TransactionalRequestResult result) {
+            this.result = result;
+        }
+
+        @Override
+        public void onComplete(ClientResponse response) {
+            if (response.requestHeader().correlationId() != inFlightRequestCorrelationId)
+                throw new IllegalStateException("Cannot have more than one transactional request in flight.");
```

Comment:
Exceptions thrown from the completion handler will be caught by `NetworkClient`. Is the intent to propagate this back to the user? We probably need to set the error in the result and move error checking outside.